### DISCUSSION TITLE
How can i get access to the error returned by the handler using a middleware?

### DISCUSSION THREAD

**Discussion by barata0** [danielgtaylor/huma#663] [General]
Let´s suppose I want to log every error returned by the handlers
I believe the best way to do this is using a middleware and if there is an error log it.

Best I could think about is this:

```if ctx.Status() >= 400 {
		log.Printf("Error: %v", ctx.Context().Err())
	}
``` 

But ctx.Context.Err is nil.

**Comment by danielgtaylor** [comment]
@barata0 it depends on what you want to log exactly. I would suggest either a middleware or a transformer:

- Log HTTP status code, headers, response body, etc via a [middleware](https://huma.rocks/features/middleware/). You can wrap `http.ResponseWriter` or `huma.Context` to store the status/headers/body and then read it after the operation completes.
- Get the `error` instance via a [transformer](https://huma.rocks/features/response-transformers/) before it is serialized to HTTP. You will have to check if the value is an `error` instance.

Another possibility is to use your own register function, something like:

```go
func MyRegister[I, O any](api huma.API, op huma.Operation, handler func(context.Context, *I) (*O, error)) {
	huma.Register(api, op, func(ctx context.Context, input *I) (*O, error) {
		result, err := handler(ctx, input)
		if err != nil {
			// Do something here!
		}
		return result, err
	})
}
```

**Reply by shasderias** [reply]
Sorry for the cross-posts, I saw several variants of this question being asked and spent quite a bit of time trying to solve it for myself. It seems to me overriding `huma.NewErrorWithContext` although not possibly intended for this purpose, basically solves the problem.

ISTM that in `huma.NewErrorWithContext`, you have access to a `huma.Context` and the error returned by the handler, which should be good enough for most usecases (logging errors, hiding errors, transforming errors, etc.).

My original reply is here, and it comes with a little code snippet that demonstrates what I'm trying to say.

https://github.com/danielgtaylor/huma/issues/482#issuecomment-2851172185
